### ISSUE TITLE
Ability to define projection attributes in `@indexes`

### ISSUE THREAD

**Issue by filmaj** [architect/architect#1083]
**Is your feature request related to a problem? Please describe.**
Currently any index defined in `app.arc` is set [to project all attributes into the index](https://github.com/architect/package/blob/master/src/visitors/indexes/index.js#L22). Since DDB charges for storage, this can get expensive for large tables, especially in a single-table configuration where you most likely only want a subset of attributes showing up in each index.

**Describe the solution you'd like**
Would be cool to be able to specify one of the three [index attribute projection options AWS provides](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/GSI.html#GSI.Projections):

- `ALL`, the current option, which would remain the default
- `KEYS_ONLY`, which would project only the base table partition and sort key plus the index keys
- `INCLUDE`, where a list of specific attribute names could be provided.

Open to all options on how this would look like. One idea would be to provide a `ProjectionType` / `project` property under the index name. Some examples:

```
@indexes
accounts
  email *String
  createdAt **String
  ProjectionType ALL # this is default
```

```
@tables
accounts
  username *String

@indexes
accounts
  email *String
  createdAt **String
  project KEYS_ONLY # projects only the base table and index keys (so in this example username, email and createdAt)
```

```
@indexes
@tables
accounts
  username *String

accounts
  email *String
  createdAt **String
  project homepageUrl avatarUrl # projects the base table and index keys plus the attributes specified after `project` (so username, email, createdAt, homepageUrl and avatarUrl)
```

**Comment by filmaj** [comment]
Hey @ryanblock @brianleroux I am thinking of taking a stab at implementing this. Do you have any feedback on the API / manifest additions I suggest above?

**Comment by brianleroux** [comment]
I like `project`

**Comment by filmaj** [comment]
Random thought here: what if someone defined a `project` attribute in one of their `@tables`, and then wanted to create an index on that attribute?

To be fair, this issue already exists with named index support ([which we support via the `name` special property](https://arc.codes/docs/en/reference/app.arc/indexes)), as well as the options `stream true` and `encrypt mysecretkey` under `@tables`.

Last but not least: we've done a rather poor job with regards to the API symmetry in tables and indexes, as our special-case property names for these pragmas include the list `name`, `stream`, `encrypt` and `PointInTimeRecovery` 😂 . Oh boy 🙈

**Comment by filmaj** [comment]
Immediate follow-up to my own question because I guess that's how my brain works: I suppose we can inspect the value of the `project` property under `@indexes`, and if is a key-like value (a value that contains `*` or `**`), then we can assume we are dealing with the `project`-the-attribute-name and not `project`-the-projection-definition.

**Comment by filmaj** [comment]
Poking away at implementing this, and looks like I will need to modify the inventory's `indexes` config to support this.

@ryanblock thoughts on adding a `projection` attribute to the indexes inventory, which can have one of the following values?

- `ALL` (default)
- `KEYS_ONLY`
- an array of strings, which would be assumed to be DynamoDB item attribute names

Also related to this: I want to look into adding additional validation to inventory to catch/transparently smooth over certain edge cases with indexes such as defining an index key attribute called `name` _and_ using the `name` property to give the index a name (think this can be done by inspecting the value of the `name` param provided).

**Comment by ryanblock** [comment]
Looking at the following cases:
```
@tables
table-1
  name *String
  stream **Number
table-2
  encrypt *String
  PointInTimeRecovery **String
table-3
  whatever *String
  projection **String

@indexes
table-3
  projection *String
  projection something somethingElse idkWhatev
```

As you mentioned above: `table-1` + `table-2` are accommodated by the `*` + `**` syntax; if it's not a primary/secondary key, treat it as a setting. In the case of `table-3`'s index, same thing applies!

I personally prefer `projection` to `project` (just because `project` as a word is kind of overloaded).

> thoughts on adding a projection attribute to the indexes inventory, which can have one of the following values? `ALL` (default), `KEYS_ONLY`, an array of strings

Sounds excellent, can we lowcase and shorten, so just `all` + `keys` + strings?

> we've done a rather poor job with regards to the API symmetry in tables and indexes, as our special-case property names for these pragmas include ... `PointInTimeRecovery`

Yeah we should continue supporting that casing but change it to `pitr` and document as such.

**Comment by filmaj** [comment]
> I personally prefer `projection` to `project` (just because project as a word is kind of overloaded).

Sounds good to me, will aim for that.

> Sounds excellent, can we lowcase and shorten, so just `all` + `keys` + strings?

👍 

> Yeah we should continue supporting that casing but change it to `pitr` and document as such.

Sounds good to me, I'll spin this one out to its own issue though as it's sort of its own thing.

**Comment by filmaj** [comment]
I've finally started tackling this, first is there's a PR up for review in inventory: https://github.com/architect/inventory/pull/27

**Comment by ryanblock** [comment]
Just a note to say as soon as we change `project` to `projection` I think this can ship (with `package` integration ofc)!

**Comment by ryanblock** [comment]
Shipped in 10.9, bummed this took so long to get to but glad it's out and available for everyone to use!

**Comment by filmaj** [comment]
Nice!!!
